### PR TITLE
Evict orphans which are included or precluded by accepted blocks.

### PR DESCRIPTION
**Pull request by gmaxwell**
This eliminates the primary leak that causes the orphan map to
 always grow to its maximum size.

This does not go so far as to attempt to connect orphans made
 connectable by a new block.

Keeping the orphan map less full helps improve the reliability
 of relaying chains of transactions.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -2365,6 +2375,14 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
             prevheights.resize(tx.vin.size());
             for (size_t j = 0; j < tx.vin.size(); j++) {
                 prevheights[j] = view.AccessCoins(tx.vin[j].prevout.hash)->nHeight;
+                // Which orphan pool entries must we evict?
+                map<COutPoint, set<uint256> >::iterator itByOutPoint = mapOrphanTransactionsByOutPoint.find(tx.vin[j].prevout);
+                if (itByOutPoint != mapOrphanTransactionsByOutPoint.end()) {
+                    for (set<uint256>::iterator mi = itByOutPoint->second.begin(); mi != itByOutPoint->second.end(); ++mi) {
+                             const uint256& orphanHash = *mi;
```
Comment (pstratem) [inline src/main.cpp]: can you fix the whitespace here

**Comment by pstratem** [comment]
ACK 70b5f60c761f56e0e84583e77d20a81db3fe1424

modulo whitespace nit

**Comment by gmaxwell** [comment]
So, I'm not super happy with the behavior-- the issue is that it removes included or conflicted orphans, but if those orphans themselves have orphaned children, those won't get removed.   The behavior I'm seeing in testing is that it initially removes many transactions but over time seems to remove fewer or fewer, and I think it's because the orphanmap gets full of double-orphans that were conflicted.

I'm not super keen on the performance implications of recursively eliminating there.

**Comment by pstratem** [comment]
@gmaxwell Agreed this behavior is not ideal, but this is an improvement.

File: src/main.cpp [hunk 2]
```diff
@@ -94,6 +94,7 @@ struct COrphanTx {
 };
 map<uint256, COrphanTx> mapOrphanTransactions GUARDED_BY(cs_main);
 map<uint256, set<uint256> > mapOrphanTransactionsByPrev GUARDED_BY(cs_main);
```
Comment (sipa) [inline src/main.cpp]: It would be more efficient to store map<uint256, COrphanTx>::iterators instead of uint256's

**Comment by sipa** [comment]
@gmaxwell Here is a commit that switches the mapOrphansByPrev entirely to be by-COutPoint: https://github.com/sipa/bitcoin/tree/reworkorphans

File: src/main.cpp [hunk 3]
```diff
@@ -689,6 +690,26 @@ void EraseOrphansFor(NodeId peer)
 unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
     unsigned int nEvicted = 0;
+    static int64_t nNextSweep;
+    int64_t nNow = GetTime();
+    if (nNextSweep <= nNow) {
+        // Sweep out expired orphan pool entries:
+        int nErased = 0;
+        int64_t nMinExpTime = nNow + 15 * 60;
```
Comment (sipa) [inline src/main.cpp]: Can you turn this into a named constant?

**Comment by sipa** [comment]
utACK with nit

**Comment by gmaxwell** [comment]
@sipa Nit picked.

**Comment by laanwj** [comment]
utACK 54326a6

**Pull request merged**